### PR TITLE
Add provision to convert `ArraySection_t` to `DoLoop` 

### LINKED ISSUE #1318: `ArraySection_t` supports string slicing only.
```fortran
subroutine vecfcn(n)
    use iso_fortran_env, only: wp => real64
    implicit none
    integer :: n
    real(wp) :: fvec(n)
    real(wp) :: zero = 0.0_wp
    integer :: i

    fvec(2:n) = 0.0_wp

end subroutine vecfcn
```

>LFortran

```console
$ lfortran --implicit-typing --implicit-interface  a.f90 --show-llvm
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  Binary file "/home/pranavchiku/lfortran/src/bin/lfortran", in _start()
  File "./csu/../csu/libc-start.c", line 392, in __libc_start_main_impl()
  File "./csu/../sysdeps/nptl/libc_start_call_main.h", line 58, in __libc_start_call_main()
  File "/home/pranavchiku/lfortran/src/bin/lfortran.cpp", line 1862, in ??
    return emit_llvm(arg_file, lfortran_pass_manager,
  File "/home/pranavchiku/lfortran/src/bin/lfortran.cpp", line 775, in ??
    = fe.get_llvm(input, lm, pass_manager, diagnostics);
  File "/home/pranavchiku/lfortran/src/lfortran/fortran_evaluator.cpp", line 287, in LCompilers::FortranEvaluator::get_llvm(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, LCompilers::LocationManager&, LCompilers::PassManager&, LCompilers::diag::Diagnostics&)
    Result<std::unique_ptr<LLVMModule>> res = get_llvm2(code, lm, pass_manager, diagnostics);
  File "/home/pranavchiku/lfortran/src/lfortran/fortran_evaluator.cpp", line 309, in LCompilers::FortranEvaluator::get_llvm2(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, LCompilers::LocationManager&, LCompilers::PassManager&, LCompilers::diag::Diagnostics&)
    diagnostics, lm.files.back().in_filename);
  File "/home/pranavchiku/lfortran/src/lfortran/fortran_evaluator.cpp", line 342, in LCompilers::FortranEvaluator::get_llvm3(LCompilers::ASR::TranslationUnit_t&, LCompilers::PassManager&, LCompilers::diag::Diagnostics&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    compiler_options, run_fn, infile);
  File "/home/pranavchiku/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 6937, in LCompilers::asr_to_llvm(LCompilers::ASR::TranslationUnit_t&, LCompilers::diag::Diagnostics&, llvm::LLVMContext&, Allocator&, LCompilers::PassManager&, LCompilers::CompilerOptions&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    v.visit_asr((ASR::asr_t&)asr);
  File "/home/pranavchiku/lfortran/src/libasr/../libasr/asr.h", line 4473, in LCompilers::ASR::BaseVisitor<LCompilers::ASRToLLVMVisitor>::visit_asr(LCompilers::ASR::asr_t const&)
    void visit_asr(const asr_t &b) { visit_asr_t(b, self()); }
  File "/home/pranavchiku/lfortran/src/libasr/../libasr/asr.h", line 4449, in ??
    case asrType::unit: { v.visit_unit((const unit_t &)x); return; }
  File "/home/pranavchiku/lfortran/src/libasr/../libasr/asr.h", line 4474, in LCompilers::ASR::BaseVisitor<LCompilers::ASRToLLVMVisitor>::visit_unit(LCompilers::ASR::unit_t const&)
    void visit_unit(const unit_t &b) { visit_unit_t(b, self()); }
  File "/home/pranavchiku/lfortran/src/libasr/../libasr/asr.h", line 4207, in ??
    case unitType::TranslationUnit: { v.visit_TranslationUnit((const TranslationUnit_t &)x); return; }
  File "/home/pranavchiku/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 1280, in LCompilers::ASRToLLVMVisitor::visit_TranslationUnit(LCompilers::ASR::TranslationUnit_t const&)
    visit_symbol(*item.second);
  File "/home/pranavchiku/lfortran/src/libasr/../libasr/asr.h", line 4476, in LCompilers::ASR::BaseVisitor<LCompilers::ASRToLLVMVisitor>::visit_symbol(LCompilers::ASR::symbol_t const&)
    void visit_symbol(const symbol_t &b) { visit_symbol_t(b, self()); }
  File "/home/pranavchiku/lfortran/src/libasr/../libasr/asr.h", line 4217, in ??
    case symbolType::Function: { v.visit_Function((const Function_t &)x); return; }
  File "/home/pranavchiku/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 3527, in LCompilers::ASRToLLVMVisitor::visit_Function(LCompilers::ASR::Function_t const&)
    generate_function(x);
  File "/home/pranavchiku/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 3878, in LCompilers::ASRToLLVMVisitor::generate_function(LCompilers::ASR::Function_t const&)
    this->visit_stmt(*x.m_body[i]);
  File "/home/pranavchiku/lfortran/src/libasr/../libasr/asr.h", line 4491, in LCompilers::ASR::BaseVisitor<LCompilers::ASRToLLVMVisitor>::visit_stmt(LCompilers::ASR::stmt_t const&)
    void visit_stmt(const stmt_t &b) { visit_stmt_t(b, self()); }
  File "/home/pranavchiku/lfortran/src/libasr/../libasr/asr.h", line 4238, in ??
    case stmtType::Assignment: { v.visit_Assignment((const Assignment_t &)x); return; }
  File "/home/pranavchiku/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 4229, in LCompilers::ASRToLLVMVisitor::visit_Assignment(LCompilers::ASR::Assignment_t const&)
    this->visit_expr(*x.m_target);
  File "/home/pranavchiku/lfortran/src/libasr/../libasr/asr.h", line 4536, in LCompilers::ASR::BaseVisitor<LCompilers::ASRToLLVMVisitor>::visit_expr(LCompilers::ASR::expr_t const&)
    void visit_expr(const expr_t &b) { visit_expr_t(b, self()); }
  File "/home/pranavchiku/lfortran/src/libasr/../libasr/asr.h", line 4336, in ??
    case exprType::ArraySection: { v.visit_ArraySection((const ArraySection_t &)x); return; }
AssertFailed: ASR::is_a<ASR::Character_t>(*ASRUtils::expr_type(x.m_v)) && n_dims == 0
```

### PR DESCRIPTION
**Pull request by Pranavchiku**
Fixes #1318 and #1316.

### CODE REVIEW

File: src/libasr/pass/implied_do_loops.cpp [hunk 1]
```diff
@@ -301,7 +301,47 @@ class ImpliedDoLoopVisitor : public PassUtils::PassVisitor<ImpliedDoLoopVisitor>
...
+                head.m_increment = nullptr;
+                head.loc = head.m_v->base.loc;
+                Vec<ASR::stmt_t*> doloop_body;
+                doloop_body.reserve(al, 1);
+                if( doloop == nullptr ) {
+                    ASR::expr_t* ref = PassUtils::create_array_ref(target, idx_vars, al);
+                    ASR::stmt_t* assign = ASRUtils::STMT(ASR::make_Assignment_t(al, x.base.base.loc, ref, x.m_value, nullptr));
+                    doloop_body.push_back(al, assign);
+                } else {
+                doloop_body.push_back(al, doloop);
```
Comment (czgdp1807) [inline src/libasr/pass/implied_do_loops.cpp:344]: This is not the right place to add this transformation. How urgent is this feature for you?

Comment (Pranavchiku) [reply src/libasr/pass/implied_do_loops.cpp:344]: > This is not the right place to add this transformation.

Yeah, I was wondering the same but then I came across a similar implementation just above this else-if statement.

> How urgent is this feature for you?

Will need this for the compilation of tests of modern minpack, but no so urgent as workaround is fairly simple.

Comment (czgdp1807) [reply src/libasr/pass/implied_do_loops.cpp:344]: Okay. I will complete revamping of `array_op.cpp` on Tuesday and Wednesday. Then we shouldn't be needing this change at all.

Comment (Pranavchiku) [reply src/libasr/pass/implied_do_loops.cpp:344]: That is perfect! Kindly have a look at #1320 as well.

Comment (czgdp1807) [reply src/libasr/pass/implied_do_loops.cpp:344]: > then I came across a similar implementation just above this else-if statement.

Yeah. I noticed it today as well. So probably let's keep yours as well. Then after this and #1330 go in I will shift `x = 3` and `x(2:n) = 3` into `array_op.cpp`.

**Review by czgdp1807 — APPROVED** [review]

**Comment by Pranavchiku** [comment]
I will merge main and then squash it and merge.

**Review by certik — APPROVED** [review]

File: src/libasr/pass/implied_do_loops.cpp [hunk 2]
```diff
@@ -301,7 +301,47 @@ class ImpliedDoLoopVisitor : public PassUtils::PassVisitor<ImpliedDoLoopVisitor>
...
+                ASR::ttype_t* int32_type = ASRUtils::TYPE(ASR::make_Integer_t(al, target->base.loc, 4, nullptr, 0));
+                ASR::do_loop_head_t head;
+                head.m_v = idx_vars[i];
+                ASR::expr_t* start_expr = target_section->m_args[0].m_left;
+                ASR::arrayboundType bound_type = ASR::arrayboundType::LBound;
+                head.m_start = ASRUtils::EXPR(ASR::make_ArrayBound_t(al, target->base.loc, target, start_expr,
+                                int32_type, bound_type, nullptr));
+                bound_type = ASR::arrayboundType::UBound;
+                ASR::expr_t* end_expr = target_section->m_args[0].m_right;
+                head.m_end = ASRUtils::EXPR(ASR::make_ArrayBound_t(al, target->base.loc, target, end_expr,
```
Comment (certik) [inline src/libasr/pass/implied_do_loops.cpp:333]: This line and the previous one are wrong, the `end_expr` should not be there I think, rather that should be the "value" at the end.

Comment (certik) [reply src/libasr/pass/implied_do_loops.cpp:333]: Fixed in https://github.com/lfortran/lfortran/pull/1337.

Comment (certik) [reply src/libasr/pass/implied_do_loops.cpp:333]: (Another bug here is that ArrayBound should not be used at all, also fixed in #1337.)

Comment (Pranavchiku) [reply src/libasr/pass/implied_do_loops.cpp:333]: Ahh apologies for that and thanks for fixes!

Comment (certik) [reply src/libasr/pass/implied_do_loops.cpp:333]: No worries, thanks for implementing the initial code.

**Pull request merged**